assets/src/components/ladder.tsx

```diff
@@ -0,0 +1,190 @@
+    ? LadderDirection.OneToZero
+    : LadderDirection.ZeroToOne
+
+const HEIGHT = 500
+const WIDTH = 180
+const CENTER_TO_LINE = 40 // x-distance between the center of the ladder and the center of the line
+const CENTER_TO_VEHICLE = 60 // x-distance between the center of the ladder and the center of the vehicles
+const MARGIN_TOP_BOTTOM = 40 // space between the top of the route and the top of the viewbox
+
+enum VehicleDirection {
+  Up,
+  Down,
+}
+
+const Timepoint = ({ timepoint, cy }: { timepoint: Timepoint; cy: number }) => (
+  <>
+    <circle
+      className="m-ladder__stop-circle"
+      cx={-CENTER_TO_LINE}
+      cy={cy}
+      r="3"
+    />
+    <circle
+      className="m-ladder__stop-circle"
+      cx={CENTER_TO_LINE}
+      cy={cy}
+      r="3"
+    />
+    <text
+      className="m-ladder__timepoint-name"
+      x="0"
+      y={cy}
+      textAnchor="middle"
+      dominantBaseline="middle"
+    >
+      {timepoint.id}
+    </text>
+  </>
+)
+
+interface VehicleProps {
+  vehicle: Vehicle
+  timepoints: Timepoint[]
+  timepointSpacingY: number
+  ladderDirection: LadderDirection
+}
+
+const Vehicle = ({
+  vehicle,
+  timepoints,
+  timepointSpacingY,
+  ladderDirection,
+}: VehicleProps) => {
+  const vehicleDirection: VehicleDirection =
+    (vehicle.direction_id === 1) ===
+    (ladderDirection === LadderDirection.ZeroToOne)
+      ? VehicleDirection.Down
+      : VehicleDirection.Up
+  const x =
+    vehicleDirection === VehicleDirection.Up
+      ? CENTER_TO_VEHICLE
+      : -CENTER_TO_VEHICLE
+  const y =
+    yForVehicle(vehicle, timepoints, timepointSpacingY, vehicleDirection) || -10
+  const textY = vehicleDirection === VehicleDirection.Up ? y + 18 : y - 15
+  return (
+    <>
+      <Triangle x={x} y={y} vehicleDirection={vehicleDirection} />
+      <text
+        className="m-ladder__label"
+        x={x}
+        y={textY}
+        textAnchor="middle"
+        dominantBaseline="middle"
+      >
+        {vehicle.label}
+      </text>
+    </>
+  )
+}
+
+const Triangle = ({
+  x,
+  y,
+  vehicleDirection,
+}: {
+  x: number
+  y: number
+  vehicleDirection: VehicleDirection
+}) => {
+  const points =
+    vehicleDirection === VehicleDirection.Up
+      ? [[x, y - 10], [x - 15, y + 10], [x + 15, y + 10]]
+      : [[x, y + 10], [x + 15, y - 10], [x - 15, y - 10]]
+  const pointsString = points.map(xy => xy.join(",")).join(" ")
+  return <polygon points={pointsString} fill="black" />
+}
+
+const yForVehicle = (
+  vehicle: Vehicle,
+  timepoints: Timepoint[],
+  timepointSpacingY: number,
+  direction: VehicleDirection
+): number | null => {
+  const timepoint_status = vehicle.timepoint_status
+  if (timepoint_status) {
+    const timepoint_id: TimepointId | null = timepoint_status.timepoint_id
+    if (timepoint_id) {
+      const timepointIndex = timepoints.findIndex(
+        timepoint => timepoint.id === timepoint_id
+      )
+      if (timepointIndex !== -1) {
+        const indexFraction =
```

Comment:
Is this simply undoing work that we did on the server to format this as a percentage? If that's true would it be better if the server just sent a fraction in the first place?

Reply:
Yeah, good point. I'll do that.